### PR TITLE
Fix #23853: AI Governance and Compliance Framework for AI Applications

### LINKED ISSUE #23853: AI Governance and Compliance Framework for AI Applications
## Overview

  Implement comprehensive AI governance and regulatory compliance tracking capabilities within OpenMetadata to help organizations manage AI applications according to various regulatory frameworks including EU AI Act, NIST AI RMF, and other international standards.
 
## Context

  Organizations deploying AI applications need to comply with multiple regulatory frameworks across different jurisdictions. The EU AI Act, NIST AI Risk Management Framework, and other international regulations require detailed documentation, risk assessments, and ongoing compliance tracking. OpenMetadata should provide a unified platform for cataloging AI applications with built-in governance and compliance management.

  ## Requirements

  ### Reusable AI Compliance Type Specification

  Create a generic aiCompliance.json type that can be reused across multiple entities (AI Applications, LLM Models, MCP Servers, etc.).

  Supported Compliance Frameworks

  - EU AI Act (Regulation EU 2024/1689)
  - Singapore Model AI Governance Framework
  - Canada AIDA (Artificial Intelligence and Data Act)
  - US AI Bill of Rights
  - NIST AI Risk Management Framework
  - ISO/IEC 42001
  - UK AI Regulation
  - China AI Regulations
  - Custom frameworks

  ### Compliance Record Structure

  Each compliance record should support:
  - Framework identification: Enum identifying which regulatory framework
  - Compliance status: Compliant, PartiallyCompliant, NonCompliant, UnderReview, NotApplicable
  - Assessment metadata: Assessed by, assessment date, next review date
  - Framework-specific fields: Dedicated sections for framework-specific requirements
  - Remediation tracking: List of required remediation actions
  - Notes: Additional findings and context

  ### EU AI Act Compliance Fields

  Risk Classification

  - Classification levels: Minimal, Limited, High, Unacceptable
  - Risk rationale field for documentation

  Article 5: Prohibited AI Practices Assessment (8 boolean fields)

  - Subliminal manipulative techniques (Art 5(1)(a))
  - Exploitation of vulnerabilities (Art 5(1)(b))
  - Social scoring by public authorities (Art 5(1)(c))
  - Risk assessment based solely on profiling for predicting criminal offences (Art 5(1)(d))
  - Untargeted scraping of facial images for facial recognition databases (Art 5(1)(e))
  - Emotion recognition in workplace and education (Art 5(1)(f))
  - Biometric categorisation inferring sensitive attributes (Art 5(1)(g))
  - Real-time remote biometric identification in public spaces by law enforcement (Art 5(1)(h))

  Article 6: High-Risk AI Systems Assessment (8 boolean fields)

  - Critical infrastructure (Annex III(1))
  - Education and vocational training (Annex III(3))
  - Employment, workers management, and access to self-employment (Annex III(4))
  - Access to essential private services (Annex III(5))
  - Access to essential public services (Annex III(6))
  - Law enforcement purposes (Annex III(6))
  - Migration, asylum, and border control management (Annex III(7))
  - Administration of justice and democratic processes (Annex III(8))

  Conformity Assessment

  - Assessment required (boolean)
  - Assessment type: Internal, ThirdParty, NotRequired
  - Assessment body name
  - Certificate number
  - Certificate validity date

  Transparency Obligations (Article 50)

  - Users informed they are interacting with AI (boolean)
  - AI-generated content appropriately labeled (boolean)
  - Emotion recognition or biometric categorization disclosed (boolean)

  ### Ethical AI Assessment Framework

  Cross-framework ethical assessment covering:

  Privacy

  - Privacy level: Public, Sensitive, PersonalData

  Fairness

  - Fairness risk: Low, Medium, High
  - Bias mitigation coverage: None, Partial, Full

  Reliability & Safety

  - Reliability/safety risk: Low, Moderate, High

  Transparency

  - Transparency level: None, Partial, FullDisclosure

  Accountability

  - Has designated owner (boolean)
  - Subject to human oversight (boolean)
  - Audit trail enabled (boolean)

  Environmental Impact

  - Environmental consciousness: LowRisk, MediumRisk, HighRisk

  5. Scope and Deployment Information

  Per-compliance record (jurisdiction-relevant):
  - Scope: Internal, External, Both
  - Deployment regions: Array of geographic regions (relevant for jurisdictional compliance)
  - Affected user count: Estimated number of affected users

  6. Verification and Certification

  - Verification status (boolean)
  - Verifier name (internal auditor, external body, etc.)
  - Verification timestamp
  - Verification notes
  - Certificate URL (link to compliance documentation)

  7. Governance Metadata Enhancements

  Update GovernanceMetadata in aiApplication.json:
  - Registration status: Registered, Unregistered, PendingApproval, Approved, Rejected (for Shadow AI tracking)
  - Registration metadata: Registered by, registered at, approved by, approved at
  - Risk assessment: Risk level, risk factors, mitigations, assessor, assessment date
  - Data classification: Accesses PII (boolean), accesses sensitive data (boolean), data categories, data retention period
  - Governance policies: References to applied governance policy entities
  - AI compliance: Reference to aiCompliance type (array of compliance records)
  - Intake notes: Notes from governance intake form or review process
  - Approval comments: Comments from governance council on approval/rejection decision

### PR DESCRIPTION
**Pull request by harshach**
### Describe your changes:

Fixes #23853 

I worked on ... because ...

#
### Type of change:

- [ ] Bug fix
- [ ] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] My PR title is `Fixes <issue-number>: <short explanation>`
- [ ] I have commented on my code, particularly in hard-to-understand areas. 
- [ ] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

---

## Summary by Gitar

- **New AI governance framework:**
  - Implements comprehensive compliance tracking for AI applications supporting 9 regulatory frameworks (EU AI Act, NIST AI RMF, Singapore Model AI Governance, Canada AIDA, US AI Bill of Rights, ISO/IEC 42001, UK AI Regulation, China AI Regulations, Custom)
  - Detailed EU AI Act compliance structure with Articles 5 & 6 assessments, risk classification, conformity assessment, and transparency obligations

- **New entity types:**
  - `AIApplication` in `entity/ai/aiApplication.json` for AI systems (chatbots, agents, copilots, RAG, code generators) with governance metadata and multi-model support
  - `AIGovernancePolicy` in `entity/ai/aiGovernancePolicy.json` for policy rules including bias thresholds, data access controls, cost controls, and violation tracking
  - `LLMModel` in `entity/ai/llmModel.json` for LLM catalog with capabilities, cost tracking, and bias/performance metrics
  - `PromptTemplate` in `entity/ai/promptTemplate.json` for reusable prompt management
  - `AgentExecution` in `entity/ai/agentExecution.json` for AI agent execution logging with token usage and cost tracking
  - `LLMService` for LLM service connections (Anthropic, Azure OpenAI, Bedrock, HuggingFace, Ollama, OpenAI, VertexAI)

- **Reusable compliance type:**
  - `aiCompliance.json` in `type/aiCompliance.json` with framework-specific assessments, ethical AI evaluation (privacy, fairness, reliability, transparency, accountability, environmental impact), and verification/certification tracking

- **Backend infrastructure:**
  - 6 repository classes extending `EntityRepository` with full CRUD operations, updaters, and relationship management
  - 6 REST resource classes with mappers for entity transformations
  - 5 search index classes with Elasticsearch mappings for metadata discovery
  - Database migrations for MySQL and PostgreSQL creating entity tables with JSON storage and generated columns

- **Shadow AI detection:**
  - Registration status tracking (Registered, Unregistered, PendingApproval, Approved, Rejected) in governance metadata
  - Risk assessment with levels, factors, mitigations, and approval workflow support

---

### CODE REVIEW

File: openmetadata-spec/src/main/resources/json/schema/entity/ai/llmModel.json [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        "trainingCost": {
+          "type": "object",
+          "properties": {
+            "totalCost": {"type": "number"},
+            "currency": {"type": "string", "default": "USD"},
+            "computeHours": {"type": "number"},
+            "resourceType": {"type": "string"}
+          }
+        },
+        "hyperparameters": {
```
Comment (pmbrull) [inline openmetadata-spec/src/main/resources/json/schema/entity/ai/llmModel.json:109]: should we keep this as an open KV pair for people to define? similar to how we have it in mlmodel https://github.com/open-metadata/OpenMetadata/blob/main/openmetadata-spec/src/main/resources/json/schema/entity/data/mlmodel.json#L190

might be hard to keep tabs on every param

File: openmetadata-spec/src/main/resources/json/schema/entity/ai/llmModel.json [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+          "properties": {
+            "learningRate": {"type": "number"},
+            "batchSize": {"type": "integer"},
+            "epochs": {"type": "integer"},
+            "optimizer": {"type": "string"},
+            "warmupSteps": {"type": "integer"},
+            "weightDecay": {"type": "number"}
+          }
+        },
+        "trainingMetrics": {
```
Comment (pmbrull) [inline openmetadata-spec/src/main/resources/json/schema/entity/ai/llmModel.json:121]: same here, I guess depending on the model the metrics teams can track could be different

File: openmetadata-spec/src/main/resources/json/schema/entity/ai/aiGovernancePolicy.json [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+          "enum": ["Low", "Medium", "High", "Critical"]
+        },
+        "enabled": {
+          "type": "boolean",
+          "default": true
+        }
+      },
+      "required": ["name", "ruleType", "action"]
+    },
+    "biasThreshold": {
```
Comment (pmbrull) [inline openmetadata-spec/src/main/resources/json/schema/entity/ai/aiGovernancePolicy.json:64]: Would it make sense to define Bias Metrics with our Metrics definition, link those metrics as entityRefs to the llmModel, and then have the threshold here as an EntityReference + Threshold number? This way we're sure we are always going to be following the same definitions, and users might be able to go and add further pieces without them being coupled to the schema

File: openmetadata-spec/src/main/resources/json/schema/entity/ai/aiApplication.json [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+          "type": "string",
+          "description": "Notes from AI governance intake form or review process"
+        },
+        "approvalComments": {
+          "type": "string",
+          "description": "Comments from governance council on approval/rejection decision"
+        }
+      }
+    },
+    "biasMetrics": {
```
Comment (pmbrull) [inline openmetadata-spec/src/main/resources/json/schema/entity/ai/aiApplication.json:204]: same, looks like we're redefining bias with possibly different keys/props

File: openmetadata-spec/src/main/resources/json/schema/entity/ai/aiGovernancePolicy.json (hunk 3 above)
Comment (harshach) [reply openmetadata-spec/src/main/resources/json/schema/entity/ai/aiGovernancePolicy.json:64]: Bias metrics are generic no? teams do not exactly what they want to track here. We need to be more prescriptive than having a generic metric

File: openmetadata-spec/src/main/resources/json/schema/entity/ai/llmModel.json (hunk 1 above)
Comment (harshach) [reply openmetadata-spec/src/main/resources/json/schema/entity/ai/llmModel.json:109]: - These 6 hyperparameters are the most common ones used in LLM training (universal across PyTorch, TensorFlow, JAX)
  - Having them as structured fields provides type safety (number vs integer) and validation
  - Makes UI forms easier - we can render proper input fields with appropriate controls

File: openmetadata-spec/src/main/resources/json/schema/entity/ai/aiGovernancePolicy.json (hunk 3 above)
Comment (harshach) [reply openmetadata-spec/src/main/resources/json/schema/entity/ai/aiGovernancePolicy.json:64]: - User Experience: These are "snapshot" metrics that represent the current state. Forcing users to create separate Metric entities for every AI application's bias score adds
  significant overhead
  - Use Case: biasMetrics, performanceMetrics, etc. are NOT time-series metrics - they're current evaluation snapshots that belong to the entity
  - Governance: Having bias metrics inline makes governance checks easier - you can immediately see if an application has bias without needing to follow references
  - OpenMetadata Metric entities: These are designed for database/table metrics (like row count, column distributions), not AI evaluation metrics
  - Different semantics: EntityReference to Metrics implies these are reusable metric definitions. But "overall bias score for Application X" is specific to that application, not
  reusable
